pkg/services/sqlstore/annotation_cleanup.go

```diff
@@ -58,6 +63,11 @@ func (acs *AnnotationCleanupService) cleanAnnotations(ctx context.Context, cfg s
 	return nil
 }
 
+func (acs *AnnotationCleanupService) cleanOrphanedAnnotationTags(ctx context.Context) error {
+	sql := "DELETE FROM annotation_tag WHERE NOT EXISTS (SELECT null FROM annotation WHERE annotation_tag.annotation_id = annotation.id)"
+	return acs.executeUntilDoneOrCancelled(ctx, sql)
+}
```

Comment:
I've noticed that `annotation_tag` is only an intermediate table that relates `annotation` with `tag` entries (instead of using foreign keys). So I suspect that after this we will end up with a lot orphaned entries in the `tag` table. I think that they should be cleaned as well, WDYT?

Reply:
A valid point. We found that the number of records in the `tag` table does not increase at the same rate as the records in the `annotation_tag` table. Tags are not unique, and the same tag is used in several annotations. We only insert a new tag if it doesn't exist already ([link](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/annotation.go#L52)). 

If we delete from the `tag` table, we will need to ensure that the record is not in use by any other component. Dashboards and alert rules also use tags. I think this means the query needed to clean up orphaned tags would become considerably more complex, as we would need to check that there are no records in `database_tag` or `annotation_tag` that use a given tag as well. Very happy to be told this is not as hard as I think it is.

With that in mind, we hoped to only need to clean records from the `annotation_tag` table, because this is the part currently causing one particular hosted grafana instance to slow down significantly without manual intervention. We would like to address this issue of cleaning up records from the `tag` table in a separate PR.



@bergquist may want to add some details if I've missed something/.

Reply:
The fact that the tag entries are not unique is enough for making me change my mind and think that we should not remove them as well.